File: packages/viewer/src/spreadsheet/engine/charts.ts
import { SaxesParser, type SaxesTagNS } from "saxes";

import { parseChartReferenceFormula, resolveChartDataSource } from "./chart-references.ts";
import { rangeAddress } from "./coordinates.ts";
import type {
  SpreadsheetChart,
  SpreadsheetChartDataSource,
  SpreadsheetChartGroup,
  SpreadsheetChartInput,
  SpreadsheetChartSeries,
  SpreadsheetObjectAnchor,
  SpreadsheetScalar,
  SpreadsheetSheet,
} from "./model.ts";
import { escapeXml } from "./xml.ts";

const MAX_CHART_XML_NODES = 100_000;
const MAX_CHART_CACHE_POINTS = 1_000_000;

type ChartXmlNode = {
  attributes: Readonly<Record<string, string>>;
  children: ChartXmlNode[];
  name: string;
  text: string;
};

function parseChartXml(xml: string): ChartXmlNode | undefined {
  const roots: ChartXmlNode[] = [];
  const stack: ChartXmlNode[] = [];
  let nodeCount = 0;
  let failure: Error | undefined;
  const parser = new SaxesParser({ xmlns: true });
  parser.on("opentag", (tag: SaxesTagNS) => {
    nodeCount += 1;
    if (nodeCount > MAX_CHART_XML_NODES) {
      failure = new Error("Chart XML exceeds the node limit.");
      return;
    }
    const attributes: Record<string, string> = {};
    for (const attribute of Object.values(tag.attributes)) {
      attributes[attribute.local] = attribute.value;
      attributes[attribute.name] = attribute.value;
    }
    const node: ChartXmlNode = {
      attributes,
      children: [],
      name: tag.local,
      text: "",
    };
    const parent = stack.at(-1);
    if (parent) parent.children.push(node);
    else roots.push(node);
    stack.push(node);
  });
  parser.on("text", (text: string) => {
    const node = stack.at(-1);
    if (node) node.text += text;
  });
  parser.on("cdata", (text: string) => {
    const node = stack.at(-1);
    if (node) node.text += text;
  });
  parser.on("closetag", () => {
    stack.pop();
  });
  parser.on("error", (error: Error) => {
    failure = error;
  });
  parser.write(xml).close();
  return failure ? undefined : roots[0];
}

function child(node: ChartXmlNode | undefined, name: string): ChartXmlNode | undefined {
  return node?.children.find((candidate) => candidate.name === name);
}

function children(node: ChartXmlNode | undefined, name: string): readonly ChartXmlNode[] {
  return node?.children.filter((candidate) => candidate.name === name) ?? [];
}

function descendants(node: ChartXmlNode | undefined, name: string): readonly ChartXmlNode[] {
  if (!node) return [];
  const matches: ChartXmlNode[] = [];
  const visit = (candidate: ChartXmlNode) => {
    if (candidate.name === name) matches.push(candidate);
    for (const nested of candidate.children) visit(nested);
  };
  visit(node);
  return matches;
}

function firstText(node: ChartXmlNode | undefined, names: readonly string[]): string | undefined {
  for (const name of names) {
    const value = descendants(node, name)[0]?.text.trim();
    if (value) return value;
  }
  return undefined;
}

function cachedValues(
  container: ChartXmlNode,
  valueType: SpreadsheetChartDataSource["valueType"],
): readonly SpreadsheetScalar[] {
  const cache =
    descendants(container, valueType === "number" ? "numCache" : "strCache")[0] ??
    descendants(container, "multiLvlStrCache")[0] ??
    descendants(container, valueType === "number" ? "numLit" : "strLit")[0];
  if (!cache) return [];
  const pointCount = Number(child(cache, "ptCount")?.attributes.val ?? 0);
  const points = descendants(cache, "pt")
    .map((point) => ({
      index: Number(point.attributes.idx),
      value: child(point, "v")?.text ?? "",
    }))
    .filter(({ index }) => Number.isInteger(index) && index >= 0 && index < MAX_CHART_CACHE_POINTS);
  const length = Math.min(
    MAX_CHART_CACHE_POINTS,
    Math.max(
      Number.isInteger(pointCount) ? pointCount : 0,
      ...points.map(({ index }) => index + 1),
      0,
    ),
  );
  const values: SpreadsheetScalar[] = Array.from({ length }, () => null);
  for (const point of points) {
    if (valueType === "string") values[point.index] = point.value;
    else {
      const numeric = Number(point.value);
      values[point.index] = Number.isFinite(numeric) ? numeric : null;
    }
  }
  return values;
}

function dataSource(
  series: ChartXmlNode,
  role: "cat" | "val",
): SpreadsheetChartDataSource | undefined {
  const container = child(series, role);
  if (!container) return undefined;
  const sourceKind = ["numRef", "numLit", "strRef", "strLit", "multiLvlStrRef"].find(
    (name) => descendants(container, name).length > 0,
  );
  if (!sourceKind) return undefined;
  const valueType = sourceKind.startsWith("num") ? "number" : "string";
  const formula = descendants(container, "f")[0]?.text.trim();
  return {
    cache: cachedValues(container, valueType),
    ...(formula ? { reference: parseChartReferenceFormula(formula) } : {}),
    valueType,
  };
}

function chartType(group: ChartXmlNode): SpreadsheetChartGroup["type"] | undefined {
  if (group.name === "lineChart") return "line";
  if (group.name === "pieChart") return "pie";
  if (group.name === "barChart") {
    return child(group, "barDir")?.attributes.val === "bar" ? "bar" : "column";
  }
  return undefined;
}

export function parseChart(part: string, xml: string): SpreadsheetChart | undefined {
  const root = parseChartXml(xml);
  const plotArea = descendants(root, "plotArea")[0];
  const groups: SpreadsheetChartGroup[] = [];
  for (const groupNode of plotArea?.children ?? []) {
    const type = chartType(groupNode);
    if (!type) continue;
    const series: SpreadsheetChartSeries[] = [];
    for (const seriesNode of children(groupNode, "ser")) {
      const values = dataSource(seriesNode, "val");
      if (!values) continue;
      const categories = dataSource(seriesNode, "cat");
      const name = firstText(child(seriesNode, "tx"), ["v"]);
      series.push({
        ...(categories ? { categories } : {}),
        ...(name ? { name } : {}),
        values,
      });
    }
    if (series.length > 0) groups.push({ series, type });
  }
  if (groups.length === 0) return undefined;
  const legendNode = descendants(root, "legend")[0];
  const legendPosition = descendants(legendNode, "legendPos")[0]?.attributes.val;
  const legend = legendNode
    ? (({ b: "bottom", l: "left", r: "right", t: "top" } as const)[legendPosition as "b"] ??
      "right")
    : "none";
  const titleNode = descendants(root, "title")[0];
  const title = firstText(titleNode, ["t", "v"]);
  return {
    groups,
    id: part,
    legend,
    ...(title ? { title } : {}),
  };
}

function quotedSheet(name: string): string {
  return `'${name.replaceAll("'", "''")}'`;
}

export function chartXml(
  chart: SpreadsheetChartInput,
  sourceSheetName: (source: SpreadsheetChartInput["series"][number]["values"]) => string,
): string {
  const tag = chart.type === "line" ? "lineChart" : chart.type === "pie" ? "pieChart" : "barChart";
  const series = chart.series
    .map((item, index) => {
      const categoryFormula = `${quotedSheet(sourceSheetName(item.categories))}!${rangeAddress(item.categories.range)}`;
      const valueFormula = `${quotedSheet(sourceSheetName(item.values))}!${rangeAddress(item.values.range)}`;
      return `<c:ser><c:idx val="${index}"/><c:order val="${index}"/>${item.name ? `<c:tx><c:v>${escapeXml(item.name)}</c:v></c:tx>` : ""}<c:cat><c:strRef><c:f>${escapeXml(categoryFormula)}</c:f></c:strRef></c:cat><c:val><c:numRef><c:f>${escapeXml(valueFormula)}</c:f></c:numRef></c:val></c:ser>`;
    })
    .join("");
  const chartBody =
    tag === "barChart"
      ? `<c:barDir val="${chart.type === "bar" ? "bar" : "col"}"/><c:grouping val="clustered"/>${series}<c:axId val="48650112"/><c:axId val="48672768"/>`
      : series;
  const axes =
    tag === "pieChart"
      ? ""
      : '<c:catAx><c:axId val="48650112"/><c:scaling><c:orientation val="minMax"/></c:scaling><c:delete val="0"/><c:axPos val="b"/><c:crossAx val="48672768"/><c:crosses val="autoZero"/></c:catAx><c:valAx><c:axId val="48672768"/><c:scaling><c:orientation val="minMax"/></c:scaling><c:delete val="0"/><c:axPos val="l"/><c:crossAx val="48650112"/><c:crosses val="autoZero"/></c:valAx>';
  const legendPosition = { bottom: "b", left: "l", right: "r", top: "t" }[
    chart.legend === "none" ? "right" : chart.legend
  ];
  return `<?xml version="1.0" encoding="UTF-8" standalone="yes"?><c:chartSpace xmlns:c="http://schemas.openxmlformats.org/drawingml/2006/chart" xmlns:a="http://schemas.openxmlformats.org/drawingml/2006/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"><c:chart>${chart.title ? `<c:title><c:tx><c:rich><a:bodyPr/><a:lstStyle/><a:p><a:r><a:t>${escapeXml(chart.title)}</a:t></a:r></a:p></c:rich></c:tx></c:title>` : ""}<c:plotArea><c:layout/><c:${tag}>${chartBody}</c:${tag}>${axes}</c:plotArea>${chart.legend === "none" ? "" : `<c:legend><c:legendPos val="${legendPosition}"/><c:layout/></c:legend>`}<c:plotVisOnly val="1"/></c:chart></c:chartSpace>`;
}

function pointXml(point: {
  column: number;
  columnOffsetEmu: number;
  row: number;
  rowOffsetEmu: number;
}): string {
  return `<xdr:col>${point.column - 1}</xdr:col><xdr:colOff>${point.columnOffsetEmu}</xdr:colOff><xdr:row>${point.row - 1}</xdr:row><xdr:rowOff>${point.rowOffsetEmu}</xdr:rowOff>`;
}

export function chartAnchorXml(
  anchor: SpreadsheetObjectAnchor,
  relationshipId: string,
  objectId: number,
  name: string,
): string {
  const frame = `<xdr:graphicFrame macro=""><xdr:nvGraphicFramePr><xdr:cNvPr id="${objectId}" name="${escapeXml(name)}"/><xdr:cNvGraphicFramePr/></xdr:nvGraphicFramePr><xdr:xfrm/><a:graphic><a:graphicData uri="http://schemas.openxmlformats.org/drawingml/2006/chart"><c:chart xmlns:c="http://schemas.openxmlformats.org/drawingml/2006/chart" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" r:id="${relationshipId}"/></a:graphicData></a:graphic></xdr:graphicFrame><xdr:clientData/>`;
  if (anchor.kind === "two-cell")
    return `<xdr:twoCellAnchor><xdr:from>${pointXml(anchor.from)}</xdr:from><xdr:to>${pointXml(anchor.to)}</xdr:to>${frame}</xdr:twoCellAnchor>`;
  if (anchor.kind === "one-cell")
    return `<xdr:oneCellAnchor><xdr:from>${pointXml(anchor.from)}</xdr:from><xdr:ext cx="${anchor.size.widthEmu}" cy="${anchor.size.heightEmu}"/>${frame}</xdr:oneCellAnchor>`;
  return `<xdr:absoluteAnchor><xdr:pos x="${anchor.position.xEmu}" y="${anchor.position.yEmu}"/><xdr:ext cx="${anchor.size.widthEmu}" cy="${anchor.size.heightEmu}"/>${frame}</xdr:absoluteAnchor>`;
}

export type SpreadsheetRenderedChart = Readonly<{
  chartId: string;
  categories: readonly string[];
  legend: SpreadsheetChart["legend"];
  series: ReadonlyArray<
    Readonly<{
      name: string;
      type: SpreadsheetChartGroup["type"];
      values: readonly number[];
    }>
  >;
  title?: string;
  type: SpreadsheetChartGroup["type"];
}>;

export function renderChartModel(
  chart: SpreadsheetChart,
  sheet: SpreadsheetSheet,
  resolveSheet: (name: string) => SpreadsheetSheet | undefined = () => undefined,
): SpreadsheetRenderedChart {
  const firstGroup = chart.groups[0];
  if (!firstGroup) throw new Error("A rendered chart requires at least one chart group.");
  const first = firstGroup.series[0];
  const categories = first?.categories
    ? resolveChartDataSource(first.categories, sheet, resolveSheet).map((value) =>
        value === null ? "" : String(value),
      )
    : [];
  return {
    chartId: chart.id,
    categories,
    legend: chart.legend,
    series: chart.groups.flatMap((group) =>
      group.series.map((item, index) => ({
        name: item.name ?? `Series ${index + 1}`,
        type: group.type,
        values: resolveChartDataSource(item.values, sheet, resolveSheet).map((value) =>
          typeof value === "number" && Number.isFinite(value) ? value : 0,
        ),
      })),
    ),
    ...(chart.title ? { title: chart.title } : {}),
    type: firstGroup.type,
  };
}

File: spreadsheet-view/src/charts.rs
//! Chart part parsing and data resolution mirroring `charts.ts` and
//! `chart-references.ts`.

use crate::model::{
    ChartAreaReference, ChartDataSource, ChartGroup, ChartReference, ChartSeries, RenderedChart,
    RenderedChartSeries, Scalar,
};
use crate::xmlutil::decode_xml;

pub use crate::model::Chart;

const MAX_FORMULA_CHARACTERS: usize = 32_768;
const MAX_REFERENCE_AREAS: usize = 4_096;
const MAX_RESOLVED_POINTS: usize = 1_000_000;
/// The engine clamps cache lengths to one million points.
pub const MAX_CHART_CACHE_POINTS: f64 = 1_000_000.0;

/// A minimal XML element with local (namespace-stripped) names.
#[derive(Default)]
pub struct Node {
    pub name: String,
    pub attributes: std::collections::BTreeMap<String, String>,
    pub children: Vec<Node>,
    pub text: String,
}

impl Node {
    pub fn attr(&self, name: &str) -> Option<&str> {
        self.attributes.get(name).map(String::as_str)
    }

    pub fn child(&self, name: &str) -> Option<&Node> {
        self.children.iter().find(|candidate| candidate.name == name)
    }

    pub fn children(&self, name: &str) -> Vec<&Node> {
        self.children.iter().filter(|candidate| candidate.name == name).collect()
    }

    /// All descendants named `name` in document order.
    pub fn descendants<'a>(&'a self, name: &str) -> Vec<&'a Node> {
        let mut matches = Vec::new();
        self.collect(name, &mut matches);
        matches
    }

    fn collect<'a>(&'a self, name: &str, matches: &mut Vec<&'a Node>) {
        if self.name == name {
            matches.push(self);
        }
        for nested in &self.children {
            nested.collect(name, matches);
        }
    }
}

/// Parse XML into a namespace-stripped node tree; `None` when malformed or
/// over the engine's 100k node budget.
pub fn parse_chart_xml(source: &str) -> Option<Node> {
    use quick_xml::events::Event;

    let mut reader = quick_xml::Reader::from_str(source);
    reader.config_mut().check_end_names = true;
    let mut roots: Vec<Node> = Vec::new();
    let mut stack: Vec<Node> = Vec::new();
    let mut count = 0usize;
    let mut buffer = Vec::new();
    loop {
        let event = match reader.read_event_into(&mut buffer) {
            Ok(event) => event,
            Err(_) => return None,
        };
        match event {
            Event::Start(start) => {
                count += 1;
                if count > 100_000 {
                    return None;
                }
                let raw_name = String::from_utf8_lossy(start.name().as_ref()).into_owned();
                let mut attributes_map = std::collections::BTreeMap::new();
                for attribute in start.attributes() {
                    let attribute = match attribute {
                        Ok(attribute) => attribute,
                        Err(_) => return None,
                    };
                    let key = String::from_utf8_lossy(attribute.key.as_ref()).into_owned();
                    let local = key.rsplit(':').next().unwrap_or(&key).to_string();
                    let value = attribute
                        .normalized_value(quick_xml::XmlVersion::Implicit1_0)
                        .map(|value| value.into_owned())
                        .unwrap_or_default();
                    attributes_map.insert(local, value);
                }
                stack.push(Node {
                    name: raw_name.rsplit(':').next().unwrap_or("").to_string(),
                    attributes: attributes_map,
                    children: Vec::new(),
                    text: String::new(),
                });
            }
            Event::End(_) => {
                let finished = stack.pop()?;
                if let Some(parent) = stack.last_mut() {
                    parent.children.push(finished);
                } else {
                    roots.push(finished);
                }
            }
            Event::Empty(start) => {
                count += 1;
                if count > 100_000 {
                    return None;
                }
                let raw_name = String::from_utf8_lossy(start.name().as_ref()).into_owned();
                let mut attributes_map = std::collections::BTreeMap::new();
                for attribute in start.attributes() {
                    let attribute = match attribute {
                        Ok(attribute) => attribute,
                        Err(_) => return None,
                    };
                    let key = String::from_utf8_lossy(attribute.key.as_ref()).into_owned();
                    let local = key.rsplit(':').next().unwrap_or(&key).to_string();
                    let value = attribute
                        .normalized_value(quick_xml::XmlVersion::Implicit1_0)
                        .map(|value| value.into_owned())
                        .unwrap_or_default();
                    attributes_map.insert(local, value);
                }
                let node = Node {
                    name: raw_name.rsplit(':').next().unwrap_or("").to_string(),
                    attributes: attributes_map,
                    children: Vec::new(),
                    text: String::new(),
                };
                if let Some(parent) = stack.last_mut() {
                    parent.children.push(node);
                } else {
                    roots.push(node);
                }
            }
            Event::Text(text) => {
                let content = String::from_utf8_lossy(text.as_ref()).into_owned();
                let decoded = decode_xml(&content);
                if let Some(frame) = stack.last_mut() {
                    frame.text.push_str(&decoded);
                }
            }
            Event::CData(text) => {
                let content = String::from_utf8_lossy(text.as_ref()).into_owned();
                if let Some(frame) = stack.last_mut() {
                    frame.text.push_str(&content);
                }
            }
            Event::Eof => break,
            _ => {}
        }
        buffer.clear();
    }
    roots.into_iter().next()
}

fn first_trimmed_text(node: Option<&Node>, names: &[&str]) -> Option<String> {
    let node = node?;
    for candidate in names {
        if let Some(found) = node.descendants(candidate).first() {
            let value = found.text.trim();
            if !value.is_empty() {
                return Some(value.to_string());
            }
        }
    }
    None
}

/// The authored cache points behind a data source, padded to the declared
/// point count exactly like `cachedValues`.
fn cached_values(container: &Node, number_type: bool) -> Vec<Scalar> {
    let cache = container
        .descendants(if number_type { "numCache" } else { "strCache" })
        .into_iter()
        .next()
        .or_else(|| container.descendants("multiLvlStrCache").into_iter().next())
        .or_else(|| {
            container.descendants(if number_type { "numLit" } else { "strLit" }).into_iter().next()
        });
    let cache = match cache {
        Some(cache) => cache,
        None => return Vec::new(),
    };
    let point_count = cache
        .child("ptCount")
        .and_then(|point| point.attr("val"))
        .and_then(crate::jsnumber::parse)
        .filter(|value| value.fract() == 0.0 && *value >= 0.0)
        .map(|value| value as i64)
        .unwrap_or(0);
    let points: Vec<(i64, String)> = cache
        .descendants("pt")
        .into_iter()
        .filter_map(|point| {
            let index = point.attr("idx").and_then(crate::jsnumber::parse)?;
            if index.fract() != 0.0 || !(0.0..MAX_CHART_CACHE_POINTS).contains(&index) {
                return None;
            }
            let value = point.child("v").map(|v| v.text.clone()).unwrap_or_default();
            Some((index as i64, value))
        })
        .collect();
    let length =
        point_count.max(points.iter().map(|(index, _)| index + 1).max().unwrap_or(0)).max(0);
    let length = (length.min(MAX_CHART_CACHE_POINTS as i64)) as usize;
    let mut values: Vec<Scalar> = vec![Scalar::Null; length];
    for (index, value) in points {
        if number_type {
            values[index as usize] = crate::jsnumber::parse(&value)
                .filter(|parsed| parsed.is_finite())
                .map_or(Scalar::Null, Scalar::Number);
        } else {
            values[index as usize] = Scalar::Text(value);
        }
    }
    values
}

fn data_source(series: &Node, role: &str) -> Option<ChartDataSource> {
    let container = series.child(role)?;
    let source_kind = ["numRef", "numLit", "strRef", "strLit", "multiLvlStrRef"]
        .into_iter()
        .find(|name| !container.descendants(name).is_empty())?;
    let number_type = source_kind.starts_with("num");
    let formula = container
        .descendants("f")
        .first()
        .map(|found| found.text.trim().to_string())
        .filter(|text| !text.is_empty());
    Some(ChartDataSource {
        cache: cached_values(container, number_type),
        reference: formula.as_deref().map(parse_chart_reference_formula),
        value_type: if number_type { "number" } else { "string" },
    })
}

fn chart_group_type(group: &Node) -> Option<&'static str> {
    match group.name.as_str() {
        "lineChart" => Some("line"),
        "pieChart" => Some("pie"),
        "barChart" => {
            Some(if group.child("barDir").and_then(|dir| dir.attr("val")) == Some("bar") {
                "bar"
            } else {
                "column"
            })
        }
        _ => None,
    }
}

/// Parse a chart part; `id` mirrors the part name the engine attached.
pub fn parse_chart(id: &str, source: &str) -> Option<Chart> {
    let root = parse_chart_xml(source)?;
    let plot_area = root.descendants("plotArea").into_iter().next()?;
    let mut groups: Vec<ChartGroup> = Vec::new();
    for group_node in &plot_area.children {
        let Some(group_type) = chart_group_type(group_node) else {
            continue;
        };
        let mut series: Vec<ChartSeries> = Vec::new();
        for series_node in group_node.children("ser") {
            let Some(values) = data_source(series_node, "val") else {
                continue;
            };
            let categories = data_source(series_node, "cat");
            let name = first_trimmed_text(series_node.child("tx"), &["v"]);
            series.push(ChartSeries {
                categories,
                name: name.filter(|name| !name.is_empty()),
                values,
            });
        }
        if !series.is_empty() {
            groups.push(ChartGroup { series, r#type: group_type });
        }
    }
    if groups.is_empty() {
        return None;
    }
    let legend_node = root.descendants("legend").into_iter().next();
    let legend_position = legend_node
        .as_ref()
        .and_then(|legend| legend.descendants("legendPos").into_iter().next())
        .and_then(|position| position.attr("val").map(str::to_string));
    let legend = match legend_node {
        None => "none",
        Some(_) => match legend_position.as_deref() {
            Some("b") => "bottom",
            Some("l") => "left",
            Some("r") => "right",
            Some("t") => "top",
            _ => "right",
        },
    };
    let title = first_trimmed_text(root.descendants("title").into_iter().next(), &["t", "v"])
        .filter(|title| !title.is_empty());
    Some(Chart { groups, id: id.to_string(), legend, title })
}

// ---------------------------------------------------------------------------
// Reference formulas
// ---------------------------------------------------------------------------

fn strip_outer_parentheses(source: &str) -> String {
    let trimmed = source.trim();
    if !trimmed.starts_with('(') || !trimmed.ends_with(')') {
        return trimmed.to_string();
    }
    let characters: Vec<char> = trimmed.chars().collect();
    let mut depth = 0i32;
    let mut quoted = false;
    let mut index = 0;
    while index < characters.len() {
        let character = characters[index];
        if character == '\'' {
            if quoted && characters.get(index + 1) == Some(&'\'') {
                index += 1;
            } else {
                quoted = !quoted;
            }
            index += 1;
            continue;
        }
        if !quoted {
            if character == '(' {
                depth += 1;
            } else if character == ')' {
                depth -= 1;
            }
            if depth == 0 && index < characters.len() - 1 {
                return trimmed.to_string();
            }
            if depth < 0 {
                return trimmed.to_string();
            }
        }
        index += 1;
    }
    if depth == 0 && !quoted {
        return trimmed[1..trimmed.len() - 1].trim().to_string();
    }
    trimmed.to_string()
}

fn split_union(source: &str) -> Option<Vec<String>> {
    let characters: Vec<char> = source.chars().collect();
    let mut parts: Vec<String> = Vec::new();
    let mut start = 0usize;
    let mut depth = 0i32;
    let mut quoted = false;
    let mut index = 0;
    while index < characters.len() {
        let character = characters[index];
        if character == '\'' {
            if quoted && characters.get(index + 1) == Some(&'\'') {
                index += 1;
            } else {
                quoted = !quoted;
            }
        } else if !quoted {
            if character == '(' {
                depth += 1;
            } else if character == ')' {
                depth -= 1;
                if depth < 0 {
                    return None;
                }
            } else if character == ',' && depth == 0 {
                parts.push(characters[start..index].iter().collect::<String>().trim().to_string());
                start = index + 1;
            }
        }
        index += 1;
    }
    if quoted || depth != 0 {
        return None;
    }
    parts.push(characters[start..].iter().collect::<String>().trim().to_string());
    if parts.iter().all(|part| !part.is_empty()) { Some(parts) } else { None }
}

fn reference_separator(source: &str) -> isize {
    let characters: Vec<char> = source.chars().collect();
    let mut quoted = false;
    let mut separator: isize = -1;
    let mut index = 0;
    while index < characters.len() {
        if characters[index] == '\'' {
            if quoted && characters.get(index + 1) == Some(&'\'') {
                index += 1;
            } else {
                quoted = !quoted;
            }
        } else if !quoted && characters[index] == '!' {
            separator = index as isize;
        }
        index += 1;
    }
    if quoted { -1 } else { separator }
}

fn reference_sheet_name(source: &str) -> Option<String> {
    let trimmed = source.trim();
    if trimmed.is_empty() || trimmed.contains('[') || trimmed.contains(']') {
        return None;
    }
    if trimmed.starts_with('\'') && trimmed.ends_with('\'') {
        return Some(trimmed[1..trimmed.len() - 1].replace("''", "'"));
    }
    if !trimmed.contains('\'') && !trimmed.contains('!') {
        return Some(trimmed.to_string());
    }
    None
}

/// Parse a chart data reference like `'Sheet 1'!$A$2:$A$5` into typed areas or
/// an opaque fallback, mirroring `parseChartReferenceFormula`.
pub fn parse_chart_reference_formula(formula: &str) -> ChartReference {
    let raw = formula.trim();
    if raw.is_empty() || raw.chars().count() > MAX_FORMULA_CHARACTERS {
        return ChartReference::Opaque { formula: raw.to_string() };
    }
    let Some(operands) = split_union(&strip_outer_parentheses(raw)) else {
        return ChartReference::Opaque { formula: raw.to_string() };
    };
    if operands.len() > MAX_REFERENCE_AREAS {
        return ChartReference::Opaque { formula: raw.to_string() };
    }
    let mut areas: Vec<ChartAreaReference> = Vec::new();
    for operand in operands {
        let separator = reference_separator(&operand);
        let qualifier = if separator < 0 {
            None
        } else {
            let split_index =
                operand.char_indices().nth(separator as usize).map(|(byte_index, _)| byte_index);
            match split_index {
                Some(byte_index) => reference_sheet_name(&operand[..byte_index]),
                None => None,
            }
        };
        if separator >= 0 && qualifier.is_none() {
            return ChartReference::Opaque { formula: raw.to_string() };
        }
        let address_part: String = if separator < 0 {
            operand.clone()
        } else {
            operand
                .char_indices()
                .nth(separator as usize)
                .map(|(byte_index, _)| operand[byte_index + 1..].to_string())
                .unwrap_or_default()
        };
        let address = address_part.replace('$', "");
        let Ok(range) = crate::coordinates::parse_range_address(&address) else {
            return ChartReference::Opaque { formula: raw.to_string() };
        };
        areas.push(ChartAreaReference { range, sheet_name: qualifier });
    }
    ChartReference::Areas { areas, formula: raw.to_string() }
}

// ---------------------------------------------------------------------------
// Rendering
// ---------------------------------------------------------------------------

/// A resolved view of one sheet used during chart rendering.
pub struct SheetView {
    /// Effective values: formula cells expose their cached result.
    pub cells: std::collections::BTreeMap<(u32, u32), Scalar>,
    pub id: String,
    pub name: String,
}

pub fn resolve_chart_data_source(
    source: &ChartDataSource,
    current: &SheetView,
    sheets: &[SheetView],
) -> Vec<Scalar> {
    let Some(reference) = &source.reference else {
        return source.cache.clone();
    };
    let areas = match reference {
        ChartReference::Areas { areas, .. } => areas,
        ChartReference::Opaque { .. } => return source.cache.clone(),
    };
    let mut values: Vec<Scalar> = Vec::new();
    for area in areas {
        let sheet = match &area.sheet_name {
            Some(name) => find_sheet(sheets, name),
            None => Some(current),
        };
        let Some(sheet) = sheet else {
            return source.cache.clone();
        };
        let count = (area.range.bottom - area.range.top + 1) as usize
            * (area.range.right - area.range.left + 1) as usize;
        if values.len() + count > MAX_RESOLVED_POINTS {
            return source.cache.clone();
        }
        for row in area.range.top..=area.range.bottom {
            for column in area.range.left..=area.range.right {
                values.push(sheet.cells.get(&(row, column)).cloned().unwrap_or(Scalar::Null));
            }
        }
    }
    values
}

/// Resolution matches a sheet by id or exact name, mirroring the engine.
pub fn find_sheet<'a>(sheets: &'a [SheetView], name: &str) -> Option<&'a SheetView> {
    sheets.iter().find(|sheet| sheet.id == name || sheet.name == name)
}

/// Resolve a parsed chart to its renderable form, mirroring
/// `renderChartModel`.
pub fn render_chart_model(
    chart: &Chart,
    sheet: &SheetView,
    sheets: &[SheetView],
) -> Option<RenderedChart> {
    let first_group = chart.groups.first()?;
    let first = first_group.series.first();
    let categories = first
        .and_then(|series| series.categories.as_ref())
        .map(|source| {
            resolve_chart_data_source(source, sheet, sheets)
                .into_iter()
                .map(|value| match value {
                    Scalar::Null => String::new(),
                    Scalar::Bool(value) => if value { "true" } else { "false" }.to_string(),
                    Scalar::Text(text) => text,
                    Scalar::Number(number) => crate::jsnumber::to_js_string(number),
                })
                .collect::<Vec<_>>()
        })
        .unwrap_or_default();
    let mut rendered_series: Vec<RenderedChartSeries> = Vec::new();
    for group in &chart.groups {
        for (index, item) in group.series.iter().enumerate() {
            let values = resolve_chart_data_source(&item.values, sheet, sheets)
                .into_iter()
                .map(|value| match value {
                    Scalar::Number(number) if number.is_finite() => number,
                    _ => 0.0,
                })
                .collect();
            rendered_series.push(RenderedChartSeries {
                name: item.name.clone().unwrap_or_else(|| format!("Series {}", index + 1)),
                r#type: group.r#type,
                values,
            });
        }
    }
    Some(RenderedChart {
        chart_id: chart.id.clone(),
        categories,
        legend: chart.legend,
        series: rendered_series,
        title: chart.title.clone(),
        r#type: first_group.r#type,
    })
}
